environs/bootstrap/bootstrap.go

```diff
@@ -787,13 +787,19 @@ func finalizeInstanceBootstrapConfig(
 		return errors.Annotate(err, "encoding default controller cert to pem")
 	}
 
+	agentVersion, has := cfg.AgentVersion()
+	if !has {
+		return errors.New("finalising instance bootstrap config, agent version not set on model config")
+	}
+
 	icfg.Bootstrap.StateServingInfo = controller.StateServingInfo{
 		StatePort:    controllerCfg.StatePort(),
 		APIPort:      controllerCfg.APIPort(),
 		Cert:         string(cert),
 		PrivateKey:   string(key),
 		CAPrivateKey: args.CAPrivateKey,
 	}
+	icfg.Bootstrap.StateInitializationParams.AgentVersion = agentVersion
```

Comment:
So why do we even need to set agent version here. It's bootstrap so it must be equal to the CurrentVersion. Can't we just not pass it through and avoid the checks and ambiguity.

Reply:
Technically yes but the client is still making all the decisions in this case. I hate it and want to get rid of it. @hpidcock work will get it out of the client. At the moment the goal is to remove it from model config and make the process more robust.

I will add it to our tech debt register for this.

Reply:
Can we also add TODOs in the code where we need to make the changes, as there's a few places.

Reply:
I have added it to the risk register.